src/core/lib/resource_quota/arena.h

```diff
@@ -88,6 +131,37 @@ class Arena {
     return &p->t;
   }
 
+  class PooledDeleter {
+   public:
+    explicit PooledDeleter(Arena* arena) : arena_(arena) {}
+    PooledDeleter() = default;
+    template <typename T>
+    void operator()(T* p) {
+      // TODO(ctiller): promise based filter hijacks ownership of some pointers
+      // to make them appear as PoolPtr without really transferring ownership,
+      // by setting the arena to nullptr.
+      // This is a transitional hack and should be removed once promise based
+      // filter is removed.
+      if (arena_ != nullptr) arena_->DeletePooled(p);
+    }
+
+   private:
+    Arena* arena_;
+  };
+
+  template <typename T>
+  using PoolPtr = std::unique_ptr<T, PooledDeleter>;
+
+  template <typename T, typename... Args>
+  PoolPtr<T> MakePooled(Args&&... args) {
+    return PoolPtr<T>(
+        new (AllocPooled(
+            arena_detail::AllocationSizeFromObjectSize<sizeof(T)>(PoolSizes()),
+            &pools_[arena_detail::PoolFromObjectSize<sizeof(T)>(PoolSizes())]))
```

Comment:
Just for clarification, this is finding the smallest Pool which can fit the object right ?

Reply:
right

Reply:
What will happen if the object doesn't fit in any of the available pool sizes ? Could it crash/cause a run-time issue

Reply:
right now it won't compile